### PR TITLE
feat: Guided Navigator with Trial&Error Guider provider WIP

### PR DESCRIPTION
**Pull request by benjaminhuth**
This PR proposes a Guided Navigator as diskussed in #741. There is also provided a very simple Trial&Error Guider, that in each step tries all available surfaces.

The **general design** is something like:

```c++
template<typename Guider>
class GuidedNavigator
{
    Guider m_guider;

    // Navigator interface (status, target)
};

class DummyGuider
{
    template <typename prop_state_t, typename stepper_t>
    std::vector<const Surface *> operator()(const prop_state_t &, const stepper_t &);
};

```

**Points I think need discussion here are:**

* General design (templating on `Guider`-class good choice, std::vector of surfaces as return type?)
  * (Why) Must the navigator be default-constructable?
* How to set/track `currentVolume`
* Public interface of navigation state
* Initialization / abort conditions correct?

**Other things**
* There are some `TODO` flags spread in the code, which I have not thought through yet/I have not done yet.
* Part of this is also a example executable and a `CsvPropagationStepsWriter` to compare the results with the default navigator.

### CODE REVIEW

**Comment by paulgessinger** [comment]
I need to look at this.

File: Core/include/Acts/Propagator/GuidedNavigator.hpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    bool targetReached = false;
+
+    const TrackingVolume *currentVolume = nullptr;
+    const TrackingVolume *worldVolume = nullptr;
+    const Surface *targetSurface = nullptr;
+  };
+
+  // TODO Must the Navigator really be default constructable?
+  // See Propagator.hpp:247
+  GuidedNavigator() { throw_assert(false, "Don't use this constructor!"); }
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/GuidedNavigator.hpp:42]: That might be something we should get rid of, actually.

File: Core/include/Acts/Propagator/GuidedNavigator.hpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+      return;
+    }
+
+    // Navigator status always resets the current surface
+    state.navigation.currentSurface = nullptr;
+
+    // Establish the surface status
+    auto surfaceStatus = stepper.updateSurfaceStatus(
+        state.stepping, *state.navigation.navSurfaceIter->representation,
+        false);
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/GuidedNavigator.hpp:70]: Can't `navSurfaceIter` be `.end()` here?

File: Core/include/Acts/Propagator/GuidedNavigator.hpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    if (!navstate.worldVolume->inside(stepper.position(state.stepping))) {
+      navstate.navigationBreak = true;
+      navstate.currentVolume = nullptr;
+      ACTS_VERBOSE("not inside worldVolume, stop navigation");
+      return;
+    }
+
+    // Fetch new targets if there are no candidates in state
+    if (navstate.navSurfaceIter == navstate.navSurfaces.end()) {
+      // This means also we are currently on a surface
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/GuidedNavigator.hpp:115]: Does this have to be the case? Can't we have missed the last surface?

**Review by paulgessinger — COMMENTED** [review]
This looks like a good direction to me. How are you testing this? I started building a more robust navigator test setup over on my FSM navigator branch. If you think it's useful, we could try to have that run on this implementation. Right now, over there, I'm using the existing Navigator as ground truth, but if we go ahead with this, the try-all-Navigator would probably be better than this. 

We just have to make sure that any new Navigator (FSM or this one) reproduces the publicly observable state of the current Navigator.

I've discussed this with @asalzburger recently, but we need to make an effort to lock down as many edge cases that are covered (or not covered, actually) by the current Navigator in unit tests, and also document them as well as possible.

File: Core/include/Acts/Propagator/GuidedNavigator.hpp (hunk 1 above)
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/GuidedNavigator.hpp:42]: You mean the requirement that a navigator needs to be default constructable? Because at least I would say that a default constructed `Acts::Navigator` isn't really useful either...

File: Core/include/Acts/Propagator/GuidedNavigator.hpp (hunk 2 above)
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/GuidedNavigator.hpp:70]: At least from what I see not, since the `.target(...)` call does set the `navigationBreak`-flag, which causes an early return in the `status(...)` call.
The only possibility seems to me an empty surface-vector, which is checked by assertions (which are not compiled in the default build mode^^)

But yeah, one could also do an additional check and set the `navigationBreak`-flag accordingly...

File: Core/include/Acts/Propagator/GuidedNavigator.hpp (hunk 3 above)
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/GuidedNavigator.hpp:115]: Yeah, I think you are right... For this case one would have to stop the navigation I think...

**Comment by benjaminhuth** [comment]
Hi all,
I've added a second idea for a Guided Navigator. This was done by slightly modifiying the direct navigator, so that it can be supplied with a Guider-Object. This object then must implement the following two methods:

```c++
struct SomeGuider 
{
    template<typename propagator_state_t, typename stepper_t>
    void updateOnSurface(propagator_state_t &state, const stepper_t &stepper)
    {
        // Some implementation
    }

    template<typename propagator_state_t, typename stepper_t>
    void updateUnreachable(propagator_state_t &state, const stepper_t &stepper)
    {
        // Some implementation
    }
};
```

In case of the Direct navigator, both functions are `state.navigation.navSurfaceIter++`, but in case of a trial and error navigator, the `onSurface(...)` method would replace the `navSurfaces`-vector with a new one comprising the reachable surfaces sorted by distance.

Even though there maybe are still some things to sort out, I would see this as an elegant solution to not write down the whole navigation logic to often, and rely on the I think well tested direct navigator (where I literally only changed four lines of code to make this work).

You find these classes under `Examples/Run/GuidedNavigator2`.
